torchtitan/metrics.py

```diff
@@ -126,35 +162,45 @@ def _get_metrics_rank(parallel_dims: ParallelDims) -> int:
         metrics_log_rank = (world_size // pp_size) * (pp_size - 1)
     else:
         metrics_log_rank = 0
-
     return metrics_log_rank
 
 
 def build_metric_logger(
     job_config: JobConfig, parallel_dims: ParallelDims, tag: Optional[str] = None
 ):
     """
+    Args:
+        job_config: Configuration object containing metrics settings.
+        parallel_dims: Used to determine the rank to log metrics from if 'tb_config.rank_0_only=True'.
+        tag: Optional tag to prefix all metrics.
+
+    Returns:
+        MetricLogger instance configured based on the provided settings
+
     parallel_dims is used to determine the rank to log metrics from if 'tb_config.rank_0_only=True'.
     In that case, `_get_metrics_rank` will be used to calculate which rank acts as 'rank 0'. This is
     intended to allow logging from the 0th rank within the last pipeline stage group, in case pipeline
     parallelism is enabled, without forcing logging from all ranks to capture loss information.
     """
     dump_dir = job_config.job.dump_folder
-    tb_config = job_config.metrics
-    save_tb_folder = tb_config.save_tb_folder
-    # since we don't have run id, use current minute as the identifier
-    datetime_str = datetime.now().strftime("%Y%m%d-%H%M")
-    log_dir = os.path.join(dump_dir, save_tb_folder, datetime_str)
-
-    enable_tb = tb_config.enable_tensorboard
-    if enable_tb:
-        logger.info(
-            f"Metrics logging active. Tensorboard logs will be saved at {log_dir}"
-        )
-        if tb_config.rank_0_only:
-            enable_tb = torch.distributed.get_rank() == _get_metrics_rank(parallel_dims)
-        else:
-            rank_str = f"rank_{torch.distributed.get_rank()}"
-            log_dir = os.path.join(log_dir, rank_str)
+    metrics_config = job_config.metrics
+    log_dir = os.path.join(
+        dump_dir, metrics_config.save_tb_folder, datetime.now().strftime("%Y%m%d-%H%M")
+    )
+
+    enable_tb = metrics_config.enable_tensorboard
+    enable_wandb = metrics_config.enable_wandb
+    wandb_config = (
+        metrics_config.wandb_config if hasattr(metrics_config, "wandb_config") else None
+    )
+
+    if metrics_config.rank_0_only:
+        metrics_rank = _get_metrics_rank(parallel_dims)
+        is_metrics_rank = torch.distributed.get_rank() == metrics_rank
+        enable_tb = enable_tb and is_metrics_rank
+        enable_wandb = enable_wandb and is_metrics_rank
+    else:
+        rank_str = f"rank_{torch.distributed.get_rank()}"
+        log_dir = os.path.join(log_dir, rank_str)
 
-    return MetricLogger(log_dir, tag, enable_tb)
+    return MetricLogger(log_dir, tag, enable_tb, enable_wandb, wandb_config)
```

Comment:
From an extensibility perspective, it would be nicer to have this pattern:

```python
    if not should_log:
        return DummyLogger()
    if enable_wandb:
        return WandbLogger(wandb_config)
    if enable_tb:
        return TensorBoardLogger(log_dir, tag)
    raise NotImplementedError
```

With

```python

class DummyLogger:
    def log(self, *args, **kwargs):
        pass

    def close(self):
        pass


class TensorBoardLogger:
    def __init__(self, log_dir, tag):
        self.tag = tag
        self.writer = SummaryWriter(log_dir, max_queue=1000)

    def log(self, metrics: Dict[str, Any], step: int):
        for k, v in metrics.items():
            tag = k if self.tag is None else f"{self.tag}/{k}"
            self.writer.add_scalar(tag, v, step)

    def close(self):
        self.writer.close()
```

Since each class has its own separate logic, it's easier for forks to have their custom loggers with reduced conflicts. It also avoids the problem of optional types inside the classes (avoids assertions).

You could then import `wandb` inside the `WandbLogger` class to avoid importing it during startup

Reply:
I like the way this code looks but it goes back to the same issue we're discussing above around whether we should have multiple loggers in case people enable both wandb and tensorboard

Right now implicitly enabling both would enable whatever is the topmost condition

Reply:
My bad, I did not look at the existing comments extensively.

For that case, you (or a fork) could add a logger class that takes a list of loggers and simply iterates over them.

```python
    loggers = []
    if should_log:
        if enable_wandb:
            loggers.append(WandbLogger(wandb_config))
        if enable_tb:
            loggers.append(TensorBoardLogger(log_dir, tag))
    return Loggers(loggers)
```
This removes the need for `DummyLogger`, iterating over an empty list does nothing.

However, this only works well if the loggers agree to exposing the same interface under the hood, but my opinion is that this repo should stay simple and not think too much about this. From my previous experience, users don't enable two loggers at the same time.

Reply:
I think you've convinced me, 2 loggers seems clunky - incorporated your feedback